### PR TITLE
cmd/snap: sort tasks in snap debug timings output

### PR DESCRIPTION
**Pull request by stolowski**
Sort tasks in `snap debug timings ...` output by lanes and ready-time, with special handling of lane 0 tasks:
- tasks from lanes >0 are grouped by lanes and sorted by ready time.
- tasks from lane 0 are sorted by ready time and inserted before and after other lanes based on the min/max ready times of non-zero lanes.
- tasks from lane 0 with ready time inbetween non-zero lane tasks are not really expected in our system and will appear after non-zero lane tasks.

This PR also simplifies timings API by embedding task data directly in the protocol, and removing the need to query v2/change/<..> API separately (as it doesn't have lane data anyway).

### CODE REVIEW

File: daemon/api_debug.go [hunk 1]
```diff
@@ -131,7 +137,19 @@ func collectChangeTimings(st *state.State, changeID string) (map[string]*changeT
 
 	m := map[string]*changeTimings{}
 	for _, t := range chg.Tasks() {
+		lane := 0
+		// determine lowest lane number, we are only interested in a single lane here
+		for _, l := range t.Lanes() {
+			if l < lane {
```
Comment (pedronis) [inline daemon/api_debug.go]: this needs testing, afaict  lane starts = 0 so this would happen only if l is negative

File: cmd/snap/cmd_debug_timings.go [hunk 2]
```diff
@@ -92,29 +93,70 @@ func printTaskTiming(w io.Writer, t *Timing, verbose, doing bool) {
...
-		if timing.ChangeTimings[t.ID].DoingTime == 0 {
+	sort.Slice(tasks, func(i, j int) bool {
+		t1 := timings[tasks[i]]
+		t2 := timings[tasks[j]]
+		if t1.Lane != t2.Lane {
+			if t1.Lane == 0 {
+				return t1.ReadyTime.Before(minReadyTime)
+			}
+			if t2.Lane == 0 {
+				return t2.ReadyTime.After(maxReadyTime)
```
Comment (pedronis) [inline cmd/snap/cmd_debug_timings.go]: I'm not sure I see where/how:
> // - tasks from lane 0 with ready time between non-zero lane tasks are not really expected in our system and will
> //   appear after non-zero lane tasks.

is implemented. Thinking quickly I think we would want t2.ReadyTime >= minReadyTime here to implement that. But that's my quick thought:  it means anything in lane 0 after the start of other lanes goes after anything in the != 0 lanes.

This also means we don't need maxReadyTime.

File: daemon/api_debug.go (hunk 1 above)
Comment (stolowski) [reply daemon/api_debug.go]: Thanks, this was indeed wrong. Fixed and added a test.

File: cmd/snap/cmd_debug_timings.go (hunk 2 above)
Comment (stolowski) [reply cmd/snap/cmd_debug_timings.go]: Thanks, that's a good observation, I tweaked sorting function and updated respective test scenario (it would fail without the change).

File: cmd/snap/cmd_debug_timings.go [hunk 3]
```diff
@@ -92,29 +93,71 @@ func printTaskTiming(w io.Writer, t *Timing, verbose, doing bool) {
...
+	sort.Slice(tasks, func(i, j int) bool {
+		t1 := timings[tasks[i]]
+		t2 := timings[tasks[j]]
+		if t1.Lane != t2.Lane {
+			// if either t1 or t2 is from lane 0, then it comes before or after non-zero lane tasks
+			if t1.Lane == 0 {
+				return t1.ReadyTime.Before(minReadyTime)
+			}
+			if t2.Lane == 0 {
+				return t2.ReadyTime.After(minReadyTime)
```
Comment (pedronis) [inline cmd/snap/cmd_debug_timings.go]: I think both After and Before are not true if the times are equal so for that corner case we want this as !Before

File: cmd/snap/cmd_debug_timings_test.go [hunk 4]
```diff
@@ -241,35 +256,82 @@ func (s *SnapSuite) mockCmdTimingsAPI(c *C) {
...
+			"1": {Lane: 1, ReadyTime: mkTime("2019-01-21T00:00:00Z")},
+			"2": {Lane: 1, ReadyTime: mkTime("2019-01-20T00:00:00Z")},
+			"3": {Lane: 1, ReadyTime: mkTime("2019-01-16T00:00:00Z")},
+		},
+		Expected: []string{"3", "2", "1"},
+	}, {
+		// tasks in lanes 0, 1, 2 with tasks from line 0 before and after lanes 1, 2
+		ChangeTimings: map[string]main.ChangeTimings{
+			"1": {Lane: 1, ReadyTime: mkTime("2019-01-21T00:00:00Z")},
+			"8": {Lane: 0, ReadyTime: mkTime("2019-01-27T00:00:00Z")},
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_timings_test.go]: This can probably be listed after `"7"`
